internal/server/handlers.go

```diff
@@ -22,13 +24,146 @@ import (
+	}
+	addVersionHandler(a,
+		http.MethodGet, "/api/access-keys", "0.16.1",
+		route[listAccessKeysRequest_v0_16_1, *api.ListResponse[api.AccessKey]]{
+			routeSettings: defaultRouteSettingsGet,
+			handler: func(c *gin.Context, reqOld *listAccessKeysRequest_v0_16_1) (*api.ListResponse[api.AccessKey], error) {
+				req := &api.ListAccessKeysRequest{
+					UserID:            reqOld.UserID,
+					Name:              reqOld.Name,
+					ShowExpired:       reqOld.ShowExpired,
+					PaginationRequest: reqOld.PaginationRequest,
+				}
+				return a.ListAccessKeys(c, req)
+			},
+		})
+
+	type createAccessKeysRequestV0_18_0 struct {
+		UserID            uid.ID       `json:"userID"`
+		Name              string       `json:"name"`
+		TTL               api.Duration `json:"ttl"`
+		ExtensionDeadline api.Duration `json:"extensionDeadline"`
+	}
+	type createAccessKeyResponseV0_18_0 struct {
+		ID                uid.ID   `json:"id"`
+		Created           api.Time `json:"created"`
+		Name              string   `json:"name"`
+		IssuedFor         uid.ID   `json:"issuedFor"`
+		ProviderID        uid.ID   `json:"providerID"`
+		Expires           api.Time `json:"expires"`
+		ExtensionDeadline api.Time `json:"extensionDeadline"`
+		AccessKey         string   `json:"accessKey"`
+	}
+	addVersionHandler(a,
+		http.MethodPost, "/api/access-keys", "0.18.0",
+		route[createAccessKeysRequestV0_18_0, *createAccessKeyResponseV0_18_0]{
+			handler: func(c *gin.Context, reqOld *createAccessKeysRequestV0_18_0) (*createAccessKeyResponseV0_18_0, error) {
+				req := &api.CreateAccessKeyRequest{
+					UserID:            reqOld.UserID,
+					Name:              reqOld.Name,
+					Expiry:            reqOld.TTL,
+					InactivityTimeout: reqOld.ExtensionDeadline,
+				}
+				resp, err := a.CreateAccessKey(c, req)
+				if err != nil {
+					return nil, err
+				}
+				return &createAccessKeyResponseV0_18_0{
+					ID:                resp.ID,
+					Created:           resp.Created,
+					Name:              resp.Name,
+					IssuedFor:         resp.IssuedFor,
+					ProviderID:        resp.ProviderID,
+					Expires:           resp.Expires,
+					ExtensionDeadline: resp.InactivityTimeout,
+					AccessKey:         resp.AccessKey,
+				}, nil
+			},
+		})
+
+	type accessKeyV0_18_0 struct {
+		ID                uid.ID   `json:"id"`
+		Created           api.Time `json:"created"`
+		LastUsed          api.Time `json:"lastUsed"`
+		Name              string   `json:"name"`
+		IssuedForName     string   `json:"issuedForName"`
+		IssuedFor         uid.ID   `json:"issuedFor"`
+		ProviderID        uid.ID   `json:"providerID"`
+		Expires           api.Time `json:"expires"`
+		ExtensionDeadline api.Time `json:"extensionDeadline"`
+	}
+	addVersionHandler(a,
+		http.MethodGet, "/api/access-keys", "0.18.0",
+		route[api.ListAccessKeysRequest, *api.ListResponse[accessKeyV0_18_0]]{
+			handler: func(c *gin.Context, req *api.ListAccessKeysRequest) (*api.ListResponse[accessKeyV0_18_0], error) {
```

Comment:
I think the only downside to this approach is that when we change an api request or response struct (in this case `api.ListAccessKeysRequest`) we'll need to check any existing version handlers, and replace them as well.

For example, if in 0.19.0 we add a migration for `ListAccessKeysRequest` struct, then this line here will also need to change to `listAccessKeysRequestV0_19_0` since that's the version of the request that was relevant to this version.

I think we can mitigate this by changing how we store these handlers for previous versions. Instead of a single function for everything, we can store them by entity. That way when something changes it should be a shorter list of migrations to check.

Reply:
Ah, it turns out this is already a problem now. There's a migration for this endpoint from 0.16.0 that needs this change. I'll push it to this branch.